contracts/test/StakeTable.t.sol

```diff
@@ -545,3 +547,137 @@ contract StakeTable_register_Test is Test {
         // TODO
     }
 }
+
+contract StakeTableV2Test is S {
+    uint256 public newValue;
+
+    function initializeV2(uint256 _newValue) public reinitializer(2) {
+        newValue = _newValue;
+    }
+
+    function getVersion()
+        public
+        pure
+        virtual
+        override
+        returns (uint8 majorVersion, uint8 minorVersion, uint8 patchVersion)
+    {
+        return (2, 0, 0);
+    }
+}
+
+contract StakeTableUpgradeTest is Test {
+    StakeTable_register_Test stakeTableRegisterTest;
+
+    function setUp() public {
+        stakeTableRegisterTest = new StakeTable_register_Test();
+        stakeTableRegisterTest.setUp();
+    }
+
+    function test_upgrade_succeeds() public {
+        (uint8 majorVersion,,) = StakeTableV2Test(stakeTableRegisterTest.proxy()).getVersion();
+        assertEq(majorVersion, 1);
+
+        vm.startPrank(stakeTableRegisterTest.admin());
+        address proxy = stakeTableRegisterTest.proxy();
+        S(proxy).upgradeToAndCall(address(new StakeTableV2Test()), "");
+
+        (uint8 majorVersionNew,,) = StakeTableV2Test(proxy).getVersion();
+        assertEq(majorVersionNew, 2);
+
+        assertNotEq(majorVersion, majorVersionNew);
+        vm.stopPrank();
+    }
+
+    function test_upgrade_reverts_when_not_admin() public {
+        address notAdmin = makeAddr("not_admin");
+
+        (uint8 majorVersion,,) = StakeTableV2Test(stakeTableRegisterTest.proxy()).getVersion();
+        assertEq(majorVersion, 1);
+
+        vm.startPrank(notAdmin);
+        try S(stakeTableRegisterTest.proxy()).upgradeToAndCall(address(new StakeTableV2Test()), "")
```

Comment:
What's the difference of this flow vs using `expectRevert`?

Reply:
I think foundry's expectRevert() expects a direct revert from the call but since I'm interacting with the proxy, it's a delegate call and the the revert is isn't at the top level, so the expectRevert() wasn't picking it up.  I spent a lo t of time trying to figure this out. But i just realized that there's a way to circumvent this by adding 
`/// forge-config: default.allow_internal_expect_revert = true` at the top of the test function. So i'll update the tests with that as it's much simpler and easier to follow https://book.getfoundry.sh/cheatcodes/expect-revert

Reply:
ok that doesn't work so will have to leave the current solution but will remove it where possible (e.g. the initialization reverts)

Reply:
I was really confused because this all works fine in the fee contract tests but not here so something seemed off. I don't know why it doesn't work but my guess is that if written as in the PR now the expect revert only applies to the deployment of the impl contract and that doesn't revert so we don't catch the expected revert.

Reply:
thank you so so much for this! I had to remove multiple evm reads and writes from one program instruction. updated here https://github.com/EspressoSystems/espresso-network/commit/fc49aafca7104ae15dfeacbfc5c79d90767880a9